lib/Page.js

```diff
@@ -485,6 +541,87 @@ Page.prototype.insertSiteNav = function (pageData) {
     + '</div>';
 };
 
+/**
+ *  Inserts wrapper for page nav contents CSS manipulation
+ */
+Page.prototype.insertPageNavWrapper = function (pageData) {
+  if (this.validPageNavigationSpecifier()) {
+    const wrappedPageData = `<div id="${PAGE_NAV_CONENT_WRAPPER_ID}">\n${pageData}\n</div>`;
+    return wrappedPageData;
+  }
+  return pageData;
+};
+
+/**
+ *  Insert page navigation bar with headings up to headingIndexingLevel
+ *
+ *  A stack is used to maintain proper indentation levels for the headings at different heading levels.
+ */
+Page.prototype.insertPageNav = function () {
+  if (this.validPageNavigationSpecifier()) {
+    let headingHTML = '';
+    const headingStack = [];
+    Object.keys(this.navigableHeadings).forEach((key) => {
+      const currentHeadingLevel = this.navigableHeadings[key].level;
+      const currentHeadingHTML = `<a class="nav-link py-1" href="#${key}">`
+        + `${this.navigableHeadings[key].text}&#x200E;</a>\n`;
+      const nestedHeadingHTML = '<nav class="nav nav-pills flex-column my-0"'
+        + `style="margin-left: 5%; flex-wrap: nowrap;">\n${currentHeadingHTML}`;
+
+      if (headingStack.length === 0 || headingStack[headingStack.length - 1] === currentHeadingLevel) {
+        // Add heading without nesting, into headingHTML
+        headingHTML += currentHeadingHTML;
+      } else {
+        // Stack has at least 1 other heading level
+        let topOfHeadingStack = headingStack[headingStack.length - 1];
+        if (topOfHeadingStack < currentHeadingLevel) {
+          // Increase nesting level by 1
+          headingHTML += nestedHeadingHTML;
+        } else {
+          // Close any nested list with heading level higher than current
+          while (headingStack.length > 1 && topOfHeadingStack > currentHeadingLevel) {
+            headingHTML += '</nav>\n';
+            headingStack.pop();
+            topOfHeadingStack = headingStack[headingStack.length - 1];
+          }
+          if (topOfHeadingStack < currentHeadingLevel) {
+            // Increase nesting level by 1
+            headingHTML += nestedHeadingHTML;
+          } else {
+            headingHTML += currentHeadingHTML;
```

Comment:
There seems to be a mixture of determining how to partition the headings, and actual generation of the HTML code. Is it possible to separate the two? Otherwise the function gets really hard to read.

Reply:
The actual HTML generation needs to be concurrent with partitioning. It builds it from top to bottom and at each stage. I guess I can apply SLAP and bring the HTML generation to another function, similar to `formatSiteNav`?

Reply:
> I guess I can apply SLAP and bring the HTML generation to another function, similar to `formatSiteNav`?

I think that's a good place to start. We can see if there's any further improvements to be made after following that.